activiti-cloud-modeling-service/activiti-cloud-services-modeling/activiti-cloud-services-modeling-service/src/main/java/org/activiti/cloud/services/modeling/validation/process/IntermediateFlowNodeIncomingOutgoingFlowValidator.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2017-2020 Alfresco Software, Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.activiti.cloud.services.modeling.validation.process;
+
+import org.activiti.bpmn.model.EndEvent;
+import org.activiti.bpmn.model.FlowNode;
+import org.activiti.bpmn.model.StartEvent;
+import org.activiti.cloud.modeling.api.ModelValidationError;
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class IntermediateFlowNodeIncomingOutgoingFlowValidator implements FlowNodeFlowsValidator {
+
+    public static final String NO_INCOMING_FLOW_PROBLEM = "Intermediate Flow node has no incoming flow";
+    public static final String NO_INCOMING_FLOW_PROBLEM_DESCRIPTION = "Intermediate Flow node has to have an incoming flow";
+    public static final String NO_OUTGOING_FLOW_PROBLEM = "Intermediate Flow node has no outgoing flow";
+    public static final String NO_OUTGOING_FLOW_PROBLEM_DESCRIPTION = "Intermediate Flow node has to have an outgoing flow";
+    public static final String INTERMEDIATE_FLOW_VALIDATOR_NAME = "BPMN Intermediate Flow node validator";
+
+    @Override
+    public List<ModelValidationError> validate(FlowNode flowNode) {
+        List<ModelValidationError> errors = new ArrayList<>();
+
+        if (CollectionUtils.isEmpty(flowNode.getIncomingFlows())) {
+            errors.add(createModelValidationError(NO_INCOMING_FLOW_PROBLEM,
+                NO_INCOMING_FLOW_PROBLEM_DESCRIPTION,
+                INTERMEDIATE_FLOW_VALIDATOR_NAME));
+        }
+
+        if (CollectionUtils.isEmpty(flowNode.getOutgoingFlows())) {
+            errors.add(createModelValidationError(NO_OUTGOING_FLOW_PROBLEM,
+                NO_OUTGOING_FLOW_PROBLEM_DESCRIPTION,
+                INTERMEDIATE_FLOW_VALIDATOR_NAME));
+        }
+        return errors;
+    }
+
+    @Override
+    public boolean canValidate(FlowNode flowNode) {
+        return !(flowNode instanceof StartEvent) && !(flowNode instanceof EndEvent);
```

Comment:
You also need to exclude event-subprocesses (`EventSubProcess`):
![image](https://user-images.githubusercontent.com/26007058/98668557-55df8580-2350-11eb-8766-d7bfcc364b79.png)

The following XML should pass the validation. Please add an integration test for it:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn2:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn2="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" id="model-d624d23e-1082-4b19-ac5c-3ff60d3e2975" name="event-subprocess" targetNamespace="http://bpmn.io/schema/bpmn" xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd">
  <bpmn2:process id="Process_wB3JXOkJN" name="event-subprocess" isExecutable="true">
    <bpmn2:documentation />
    <bpmn2:startEvent id="Event_1">
      <bpmn2:outgoing>SequenceFlow_0yge099</bpmn2:outgoing>
    </bpmn2:startEvent>
    <bpmn2:task id="Task_0af35yr" name="Main task">
      <bpmn2:incoming>SequenceFlow_0yge099</bpmn2:incoming>
      <bpmn2:outgoing>SequenceFlow_0pv0hhn</bpmn2:outgoing>
    </bpmn2:task>
    <bpmn2:sequenceFlow id="SequenceFlow_0yge099" sourceRef="Event_1" targetRef="Task_0af35yr" />
    <bpmn2:endEvent id="EndEvent_03l0htc">
      <bpmn2:incoming>SequenceFlow_0pv0hhn</bpmn2:incoming>
    </bpmn2:endEvent>
    <bpmn2:sequenceFlow id="SequenceFlow_0pv0hhn" sourceRef="Task_0af35yr" targetRef="EndEvent_03l0htc" />
    <bpmn2:subProcess id="SubProcess_0o8kx5r" triggeredByEvent="true">
      <bpmn2:task id="Task_1lpq4jr" name="Task in subprocess">
        <bpmn2:incoming>SequenceFlow_0ep94tv</bpmn2:incoming>
        <bpmn2:outgoing>SequenceFlow_14eyqa8</bpmn2:outgoing>
      </bpmn2:task>
      <bpmn2:endEvent id="EndEvent_0hxf3ti">
        <bpmn2:incoming>SequenceFlow_14eyqa8</bpmn2:incoming>
      </bpmn2:endEvent>
      <bpmn2:sequenceFlow id="SequenceFlow_0ep94tv" sourceRef="StartEvent_0pmugnv" targetRef="Task_1lpq4jr" />
      <bpmn2:sequenceFlow id="SequenceFlow_14eyqa8" sourceRef="Task_1lpq4jr" targetRef="EndEvent_0hxf3ti" />
      <bpmn2:startEvent id="StartEvent_0pmugnv">
        <bpmn2:outgoing>SequenceFlow_0ep94tv</bpmn2:outgoing>
        <bpmn2:signalEventDefinition id="SignalEventDefinition_0l3yn3v" signalRef="Signal_0p01hxz" />
      </bpmn2:startEvent>
    </bpmn2:subProcess>
  </bpmn2:process>
  <bpmn2:signal id="Signal_0p01hxz" name="go" />
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_wB3JXOkJN">
      <bpmndi:BPMNShape id="_BPMNShape_Event_2" bpmnElement="Event_1">
        <dc:Bounds x="342" y="232" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Task_0af35yr_di" bpmnElement="Task_0af35yr">
        <dc:Bounds x="470" y="210" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_0yge099_di" bpmnElement="SequenceFlow_0yge099">
        <di:waypoint x="378" y="250" />
        <di:waypoint x="470" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="EndEvent_03l0htc_di" bpmnElement="EndEvent_03l0htc">
        <dc:Bounds x="662" y="232" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_0pv0hhn_di" bpmnElement="SequenceFlow_0pv0hhn">
        <di:waypoint x="570" y="250" />
        <di:waypoint x="662" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="Task_1lpq4jr_di" bpmnElement="Task_1lpq4jr">
        <dc:Bounds x="475" y="380" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_0ep94tv_di" bpmnElement="SequenceFlow_0ep94tv">
        <di:waypoint x="423" y="420" />
        <di:waypoint x="475" y="420" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="EndEvent_0hxf3ti_di" bpmnElement="EndEvent_0hxf3ti">
        <dc:Bounds x="607" y="402" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_14eyqa8_di" bpmnElement="SequenceFlow_14eyqa8">
        <di:waypoint x="575" y="420" />
        <di:waypoint x="607" y="420" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="SubProcess_08j2v4j_di" bpmnElement="SubProcess_0o8kx5r" isExpanded="true">
        <dc:Bounds x="345" y="330" width="350" height="200" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="StartEvent_1xhrzpm_di" bpmnElement="StartEvent_0pmugnv">
        <dc:Bounds x="387" y="402" width="36" height="36" />
      </bpmndi:BPMNShape>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn2:definitions>
```

On the other hand, embedded subprocesses (not triggered via events should have validation) and the following should fail the validation as the embeded sub-process doesn't have any incoming or outgoing flows:

![image](https://user-images.githubusercontent.com/26007058/98669340-8e339380-2351-11eb-85b9-93768ecb3d32.png)
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn2:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn2="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" id="model-679b467b-f441-48c9-bf9c-69e2c3067aff" name="embeded-sub-process" targetNamespace="http://bpmn.io/schema/bpmn" xsi:schemaLocation="http://www.omg.org/spec/BPMN/20100524/MODEL BPMN20.xsd">
  <bpmn2:process id="Process_plPmux2Yy" name="embeded-sub-process" isExecutable="true">
    <bpmn2:documentation />
    <bpmn2:startEvent id="Event_1">
      <bpmn2:outgoing>SequenceFlow_0imcv9e</bpmn2:outgoing>
    </bpmn2:startEvent>
    <bpmn2:task id="Task_1fnls1d">
      <bpmn2:incoming>SequenceFlow_0imcv9e</bpmn2:incoming>
      <bpmn2:outgoing>SequenceFlow_0y98alj</bpmn2:outgoing>
    </bpmn2:task>
    <bpmn2:sequenceFlow id="SequenceFlow_0imcv9e" sourceRef="Event_1" targetRef="Task_1fnls1d" />
    <bpmn2:endEvent id="EndEvent_0vi8yr7">
      <bpmn2:incoming>SequenceFlow_0y98alj</bpmn2:incoming>
    </bpmn2:endEvent>
    <bpmn2:sequenceFlow id="SequenceFlow_0y98alj" sourceRef="Task_1fnls1d" targetRef="EndEvent_0vi8yr7" />
    <bpmn2:subProcess id="SubProcess_1j83p8h" name="Embedded sub-process">
      <bpmn2:startEvent id="StartEvent_1dhsk7u">
        <bpmn2:outgoing>SequenceFlow_0qqat5x</bpmn2:outgoing>
      </bpmn2:startEvent>
      <bpmn2:task id="Task_0kzyhu8">
        <bpmn2:incoming>SequenceFlow_0qqat5x</bpmn2:incoming>
        <bpmn2:outgoing>SequenceFlow_1wvp9w4</bpmn2:outgoing>
      </bpmn2:task>
      <bpmn2:sequenceFlow id="SequenceFlow_0qqat5x" sourceRef="StartEvent_1dhsk7u" targetRef="Task_0kzyhu8" />
      <bpmn2:endEvent id="EndEvent_1ug8te0">
        <bpmn2:incoming>SequenceFlow_1wvp9w4</bpmn2:incoming>
      </bpmn2:endEvent>
      <bpmn2:sequenceFlow id="SequenceFlow_1wvp9w4" sourceRef="Task_0kzyhu8" targetRef="EndEvent_1ug8te0" />
    </bpmn2:subProcess>
  </bpmn2:process>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_plPmux2Yy">
      <bpmndi:BPMNShape id="_BPMNShape_Event_2" bpmnElement="Event_1">
        <dc:Bounds x="412" y="222" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Task_1fnls1d_di" bpmnElement="Task_1fnls1d">
        <dc:Bounds x="520" y="200" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_0imcv9e_di" bpmnElement="SequenceFlow_0imcv9e">
        <di:waypoint x="448" y="240" />
        <di:waypoint x="520" y="240" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="EndEvent_0vi8yr7_di" bpmnElement="EndEvent_0vi8yr7">
        <dc:Bounds x="662" y="222" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_0y98alj_di" bpmnElement="SequenceFlow_0y98alj">
        <di:waypoint x="620" y="240" />
        <di:waypoint x="662" y="240" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="SubProcess_1j83p8h_di" bpmnElement="SubProcess_1j83p8h" isExpanded="true">
        <dc:Bounds x="380" y="330" width="350" height="200" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="StartEvent_1dhsk7u_di" bpmnElement="StartEvent_1dhsk7u">
        <dc:Bounds x="422" y="412" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Task_0kzyhu8_di" bpmnElement="Task_0kzyhu8">
        <dc:Bounds x="500" y="390" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_0qqat5x_di" bpmnElement="SequenceFlow_0qqat5x">
        <di:waypoint x="458" y="430" />
        <di:waypoint x="500" y="430" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape id="EndEvent_1ug8te0_di" bpmnElement="EndEvent_1ug8te0">
        <dc:Bounds x="642" y="412" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="SequenceFlow_1wvp9w4_di" bpmnElement="SequenceFlow_1wvp9w4">
        <di:waypoint x="600" y="430" />
        <di:waypoint x="642" y="430" />
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn2:definitions>
```

Reply:
Many thanks for the help, I have added these 2 new integration checks accordingly.